### PR TITLE
Support nested exceptions and composite warning messages

### LINKED ISSUE #126: Include both message and caught error
Currently I write something like this (or some other workaround)
```julia
try
    ...
catch err
    err isa ArgumentError || rethrow(err)
    warn(logger, "message...")
    error(logger, err)
end
```

But perhaps it should be possible to write this?
```julia
error(logger, "message...", err)
```

### PR DESCRIPTION
**Pull request by rofinn**
Closes #126

### CODE REVIEW

File: src/loggers.jl [hunk 1]
```diff
@@ -460,6 +460,12 @@ let
                         end
                         throw(exc)
                     end
+
+                    function $level(logger::Logger, msg::AbstractString, exc::Exception)
+                        e = ErrorException(msg * sprint(showerror, exc))
+                        log(logger, $key, sprint(showerror, e))
```
Comment (omus) [inline src/loggers.jl]: Could we eliminate a call to `showerror`?
```julia
msg *= sprint(showerror, exc)
e = ErrorException(msg)
log(logger, $key, msg)
throw(e)
```

File: test/loggers.jl [hunk 2]
```diff
@@ -98,6 +98,21 @@
                 "I'm an error msg."
             end
             @test occursin("I'm an error msg.", String(take!(io)))
+            msg = "Failed to process stage Foo of program: "
+            @test_throws ErrorException try
+                sqrt(-1)
```
Comment (omus) [inline test/loggers.jl]:
```suggestion
                sqrt(-1)  # Throws `DomainError`
```

File: test/loggers.jl [hunk 3]
```diff
@@ -98,6 +98,21 @@
...
+            end
+            @test occursin(msg, String(take!(io)))
+
+            msg = "Foo failed, falling back to method Bar: "
+            try
+                sqrt(-1)
+            catch e
+                Memento.warn(logger, msg, e)
+            end
+            @test occursin(msg, String(take!(io)))
```
Comment (omus) [inline test/loggers.jl]: Should you also check that the output of the `DomainError` is included?

File: src/loggers.jl (hunk 1 above)
Comment (rofinn) [reply src/loggers.jl]: Sure that would work

File: src/loggers.jl [hunk 4]
```diff
@@ -460,6 +460,12 @@ let
                         end
                         throw(exc)
                     end
+
+                    function $level(logger::Logger, msg::AbstractString, exc::Exception)
+                        msg *= sprint(showerror, exc)
```
Comment (nickrobinson251) [inline src/loggers.jl]: would it be nice to be able to write
```julia
warn(getlogger(), "Woops", SomeError("Nope."))
```
(note I've "forgotten" to add a space (i.e. `"Woops "`)) 

and have it automatically render with a space as `Woops SomeError...`

Comment (rofinn) [reply src/loggers.jl]: Sure

Comment (rofinn) [reply src/loggers.jl]:
```
julia> e = DomainError(-1.0, "sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).")
DomainError(-1.0, "sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).")

julia> warn(getlogger(), "Woops", e)
[warn | root]: Woops: DomainError with -1.0:
sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
```

Comment (rofinn) [reply src/loggers.jl]: Hmmm, maybe a newline would be better.

**Pull request merged**